[develop] Adds a YAML interface for creating a Rocoto XML.

christinaholtNOAA opened the pull request:
## DESCRIPTION OF CHANGES: 

Refactors the creation of a Rocoto XML to use a very generic Jinja2 template that is flexible enough to meet the needs of various workflow configurations supported by SRW. Specifically, it allows for a completely arbitrary workflow to be created under SRW, which includes the addition of completely arbitrary tasks on top of the predefined ones. 

The workflow has been refactored to allow for the definition a workflow given specific entries in the rocoto section of the configuration files. The paradigm shifts to telling the configuration which sets of tasks to run, and removing a section when certain pre-configured tasks should not run.

The high level overview of changes include:

- Moved batch job related config variables out of config_default.yaml sections, and into the Rocoto config files for the given job. This resulted in the removal of entire sections of config_default.yaml. References to those removed sections were removed throughout SRW.
- The transition of the FV3LAM_wflow.xml template to a complete generic interpretation of YAML configs.
- The addition of parm/wflow files in groups of tasks that are commonly run together. All the Rocoto-required information for each task now lives in these yaml files.
- A modification to specifying varying forecast lengths to remove a dependence on writing all cycles in a workflow to the config file. The FCST_LEN_CYCL array is expected to correspond to hours in the day associated with a cycling freqency. For example, a list of 4 forecast lengths with a 6 hour cycling freqency indicates that the forecast lengths correspond to 0, 6, 12, and 18 UTC.
- The removal of the workflow_switches section. Whether tasks run is now based on their being defined as part of the rocoto section of the config file.
- The number of processors is now included as an environment variable passed via Rocoto instead of computed in the scripts since Rocoto holds the source of that information.
- Edits to the test config files to reflect how the test should have originally been run
- Writes a new file to the experiment directory -- rocoto_defns.yaml. It contains the rocoto section of the config that has been removed from the var_defns.yaml.

A follow-on PR will be needed to create the documentation necessary to support this change, although some documentation will be included with this PR. 

I hope to prioritize this PR as it has been, and will continue to be very difficult to support as changes are made to the existing workflow and scripts. I will work with others impacted by the change to ensure their workflow changes make it into the new workflow config files.

A demo of this tool is scheduled for this Thursday, March 16. It will be recorded.

The code has run through the comprehensive tests on Hera. Comparisons to the answers from the corresponding develop branch (up to date as of Mar 13) are underway, and being coordinated with @mkavulich. At the very least, fundamental tests should be run on each machine to test the changes necessary for each other platform.

Because there are tons of changes spanning many months of development here, I will leave the PR in a Draft state until it is sufficiently cleaned up and ready for review.



### Type of change
<!-- Please delete options that are not relevant. Add an X to check off a box. -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## TESTS CONDUCTED: 
<!-- Explicitly state what tests were run on these changes, or if any are still pending (for README or other text-only changes, just put "None required"). Make note of the compilers used, the platform/machine, and other relevant details as necessary. For more complicated changes, or those resulting in scientific changes, please be explicit! -->
<!-- Add an X to check off a box. -->

- [x] hera.intel
- [ ] orion.intel
- [ ] cheyenne.intel
- [ ] cheyenne.gnu
- [ ] gaea.intel
- [ ] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [ ] fundamental test suite
- [x] comprehensive tests (specify *which* if a subset was used)

## DEPENDENCIES:

None.

## DOCUMENTATION:

More extensive documentation is planned as a follow-on PR due to the time-sensitive nature of getting this PR prepared and merged to reduce the overhead of syncing with new changes.

## ISSUE: 
<!-- If this PR is resolving or referencing one or more issues, in this repository or elsewhere, list them here (Remember, issues must always be created before starting work on a PR branch!). For example, "Fixes issue mentioned in #123" or "Related to bug in https://github.com/ufs-community/other_repository/pull/63" -->
Addresses Issue #463 

## CHECKLIST
<!-- Add an X to check off a box. -->
- [ ] My code follows the style guidelines in the Contributor's Guide
- [ ] I have performed a self-review of my own code using the Code Reviewer's Guide
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes do not require updates to the documentation (explain).
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

christinaholtNOAA committed:
Adding XML creator.

christinaholtNOAA committed:
Decompose XML to YAML.

christinaholtNOAA committed:
Merge remote-tracking branch 'origin/develop' into rocoto_yaml

christinaholtNOAA committed:
Deprecating CYCL_HRS

Changes were made to all config files and scripts to use FIRST and LAST
cycle definitions to accept the cycle HH, and frequency will start from
those for all relevant computation.

christinaholtNOAA committed:
WIP

christinaholtNOAA committed:
Sort entries.

christinaholtNOAA committed:
Merge branch 'develop' into rocoto_yaml

christinaholtNOAA committed:
WIP

christinaholtNOAA committed:
Merge remote-tracking branch 'origin/develop' into rocoto_yaml

christinaholtNOAA committed:
Compute nprocs in yaml, not in script.

christinaholtNOAA committed:
Add machine-specific workflow flags.

christinaholtNOAA committed:
Adding a "days ago" filter.

christinaholtNOAA committed:
Removing switches from configs.

christinaholtNOAA committed:
Merge remote-tracking branch 'origin/develop' into rocoto_yaml

christinaholtNOAA committed:
Rename the default workflow definition.

christinaholtNOAA committed:
Add sections to all references to user config variables.

christinaholtNOAA committed:
WIP.

christinaholtNOAA committed:
WIP

christinaholtNOAA committed:
Works through make_lbcs, where it dies.

Need to merge develop here.

christinaholtNOAA committed:
Merge remote-tracking branch 'origin/develop' into rocoto_yaml

christinaholtNOAA committed:
Cleaning up workflow definitions

christinaholtNOAA committed:
Support for nested metatasks.

christinaholtNOAA committed:
WIP

christinaholtNOAA committed:
WIP

christinaholtNOAA committed:
Update test config files.

christinaholtNOAA committed:
Remove section from jobs

christinaholtNOAA committed:
Adjustments to XML.

christinaholtNOAA committed:
Taskvalid also follows the pattern.

christinaholtNOAA committed:
Changes to repetitive logic in get_obs scripts

christinaholtNOAA committed:
Clean up logic. Remove references to switches.

christinaholtNOAA committed:
Adjust default settings for new rocoto section.

christinaholtNOAA committed:
Clean up defaults.

christinaholtNOAA committed:
Remove memory flag for machines that can't use it.

christinaholtNOAA committed:
Cleaning up some logic.

christinaholtNOAA committed:
Merge remote-tracking branch 'origin/develop' into rocoto_yaml

christinaholtNOAA committed:
New vx task naming and cycledef for var len fcst.

christinaholtNOAA committed:
Merge remote-tracking branch 'origin/develop' into rocoto_yaml

christinaholtNOAA committed:
Remove instance of RUN_TASK variable.

christinaholtNOAA committed:
Fix typo.

christinaholtNOAA committed:
Changes needed to run comprehensive tests.

christinaholtNOAA committed:
Merge branch 'develop' into rocoto_yaml

christinaholtNOAA committed:
Cleanup after self review of PR

christinaholtNOAA committed:
Merge branch 'rocoto_yaml' of https://github.com/christinaholtNOAA/ufs-srweather-app into rocoto_yaml

christinaholtNOAA commented on tests/WE2E/test_configs/grids_extrn_mdls_suites_community/config.grid_RRFS_NA_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta.yaml:
```diff
@@ -23,15 +23,21 @@ task_get_extrn_lbcs:
+        nnodes: 12
+        ppn: 4
+        walltime: 01:00:00
+  metatask_run_ens_post:
```
This is a mistake and needs to be fixed and tested. The metatask should be nested under the tasks section.

christinaholtNOAA commented on tests/WE2E/test_configs/verification/config.MET_verification_only_vx.yaml:
```diff
@@ -15,44 +15,18 @@ workflow:
+    task_get_obs_ccpa:
+    task_get_obs_mrms:
+    task_get_obs_ndas:
+workflow_switches:
```
Remove me.

christinaholtNOAA committed:
Address review changes and moving det to mem000

christinaholtNOAA committed:
Merge remote-tracking branch 'origin/develop' into rocoto_yaml

christinaholtNOAA committed:
Changes needed after merge.

christinaholtNOAA committed:
Fix native command for fcst on Hera.

christinaholtNOAA committed:
Same change for Jet.

christinaholtNOAA committed:
Fix unittests

MichaelLueken commented on scripts/exregional_make_grid.sh:
```diff
@@ -394,11 +394,11 @@ mv_vrfy "${grid_fp_orig}" "${grid_fp}"
-# null string), check that the grid resolution contained in the variable
-# CRES set above matches the resolution appearing in the names of the
-# preexisting orography and/or surface climatology files.
+# be using (i.e. if task_make_grid task_make_surf_climo ran in the
```
Should this read:

`if task_make_grid *or* task_make_surf_climo ran in the`

MichaelLueken reviewed (approved):
@christinaholtNOAA These changes look great!  I have left a minor comment for scripts/exregional_make_grid.sh, but you don't need to make the change if you would rather not.

mkavulich commented on ush/fill_jinja_template.py:
```diff
@@ -270,6 +271,9 @@ def fill_jinja_template(argv):
     cfg = cla.config if cla.config is not None else {}
 
+    if config_dict is not None:
+        update_dict(cfg, config_dict)
```
Need to allow output to be suppressed when quiet option is active

```suggestion
        update_dict(cfg, config_dict, quiet=cla.quiet)
```

mkavulich commented on ush/setup.py:
```diff
@@ -1233,16 +1365,17 @@ def get_location(xcs, fmt, expt_cfg):
                     f"""
                            Inline post is turned on, deactivating post-processing tasks:
-                           RUN_TASK_RUN_POST = False
+                           Removing {removed_task} from task definitions
```
This produces a lot of high-priority text dumped to screen; can we either just print the name of the removed task or move the printing of the full `{removed_task}` dictionary to a debug print?

christinaholtNOAA replied:
Sure. I'll go with the 2nd option, if that's fine.

EDIT: No! I'll go with first and only have the name printed here. I didn't realize it was doing the whole dict. The name was the original intent. I thought you were referring to the sentence surrounding the name.

mkavulich commented on ush/generate_FV3LAM_wflow.py:
```diff
@@ -102,140 +102,29 @@ def generate_FV3LAM_wflow(ushdir, logfile: str = "log.generate_FV3LAM_wflow", de
-        # from the jinja template file.
+        # Call the python script to generate the experiment's XML file
         #
+        rocoto_yaml_fp = expt_config["workflow"]["ROCOTO_YAML_FP"]
```
Github won't let me do a multi-line suggestion here for some reason, but this currently prints a whole bunch of text dump to screen, even when debug=False. This logic or similar would clean things up better:

```
        rocoto_yaml_fp = expt_config["workflow"]["ROCOTO_YAML_FP"]
        args = ["-o", wflow_xml_fp,
                "-t", template_xml_fp,
                "-c", rocoto_yaml_fp ]
        if not debug:
            args.append("-q")
         try:
            fill_jinja_template(args)
         except:
```

christinaholtNOAA committed:
Addressing Mike Leuken's request

mkavulich reviewed (changes requested):
A few initial comments, still working on a full review

christinaholtNOAA committed:
Merge branch 'develop' into rocoto_yaml

christinaholtNOAA committed:
Update ush/fill_jinja_template.py

Co-authored-by: Michael Kavulich <kavulich@ucar.edu>

christinaholtNOAA committed:
Address Mike's and Eddy's feedback.

christinaholtNOAA committed:
Merge branch 'rocoto_yaml' of https://github.com/christinaholtNOAA/ufs-srweather-app into rocoto_yaml

danielabdi-noaa commented on ush/config_defaults.yaml:
```diff
@@ -963,7 +995,7 @@ nco:
   DATAROOT: '{{ OPSROOT }}/tmp'
   DCOMROOT: '{{ OPSROOT }}/dcom'
-  LOGBASEDIR: '{{ OPSROOT }}/output'
+  LOGBASEDIR: '{% if user.RUN_ENVIR == nco %}{{ [OPSROOT, output]|path_join }}{% else %}{{ [workflow.EXPTDIR, "log"]|path_join }}{% endif %}'
```
It is not writing to "output" directory when run with `run_envir=nco` mode for me, although this logic looks correct.

mkavulich replied:
I can confirm this problem. It looks like there is some error in the logic chain that results in this sort of definition in the xml file (in the community location rather than the nco location):

`  <!ENTITY LOGDIR "/scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_676/expt_dirs/156882be/nco_grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15_thompson_mynn_lam3km/log/@Y@m@d">
`

christinaholtNOAA replied:
I should be able to take a look at this by morning.

christinaholtNOAA replied:
I think `nco` should be in quotes.

danielabdi-noaa commented on parm/wflow/aqm_all.yaml:
```diff
@@ -0,0 +1,183 @@
+    maxtries: '2'
+  command: '&LOAD_MODULES_RUN_TASK_FP; "&TN_AQM_ICS;" "&JOBSDIR;/JREGIONAL_AQM_ICS"'
+  envars:
+    PREV_CYCLE_DIR: '&WARMSTART_CYCLE_DIR;'
```
WARMSTART_CYCLE_DIR does not seem to be defined.

christinaholtNOAA replied:
I think this one was resolved with my latest push.

danielabdi-noaa commented on parm/wflow/default_workflow.yaml:
```diff
@@ -0,0 +1,42 @@
+    USHdir: '{{ user.USHdir }}'
+    WORKFLOW_ID: '{{ workflow.WORKFLOW_ID }}'
+  attrs:
+    cyclethrottle: "20"
```
"taskthrottle" is missing here.

danielabdi-noaa commented on parm/wflow/default_workflow.yaml:
```diff
@@ -0,0 +1,42 @@
+    COMINgfs: '{{ platform.get("COMINgfs") }}'
+    FCST_DIR: '{{ nco.DATAROOT }}/run_fcst.{{ workflow.WORKFLOW_ID }}_@Y@m@d@H'
+    GLOBAL_VAR_DEFNS_FP: '{{ workflow.GLOBAL_VAR_DEFNS_FP }}'
+    JOBSDIR: '{{ user.JOBSdir }}'
```
Can this be "JOBSdir" to be consistent with USHdir, SCRIPTSdir etc

danielabdi-noaa commented on parm/wflow/default_workflow.yaml:
```diff
@@ -0,0 +1,42 @@
+    CCPA_OBS_DIR: '{{ platform.CCPA_OBS_DIR }}'
+    COMIN_DIR: '{% if user.RUN_ENVIR == "nco" %}{{ nco.COMIN_BASEDIR }}/{{ nco.RUN }}.@Y@m@d/@H {% else %}{{ nco.COMIN_BASEDIR }}/@Y@m@d@H{% endif %}'
+    COMINgfs: '{{ platform.get("COMINgfs") }}'
+    FCST_DIR: '{{ nco.DATAROOT }}/run_fcst.{{ workflow.WORKFLOW_ID }}_@Y@m@d@H'
```
Does this work with ensemble runs? The reason why I am asking is because the old xml has ensemble logic for DYN_DIR/PHY_DIR

christinaholtNOAA replied:
This one might need a quick followup. I will have Mike K. check results for an NCO ensemble test.

christinaholtNOAA replied:
Ahhh! This one is taken care of in setup.py around line 1108.

danielabdi-noaa commented on tests/WE2E/test_configs/grids_extrn_mdls_suites_community/config.grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp_regional_plot.yaml:
```diff
@@ -12,8 +12,9 @@ workflow:
-  RUN_TASK_PLOT_ALLVARS: true
+rocoto:
+  tasks:
+    taskgroups: '{{ ["parm/wflow/prep.yaml", "parm/wflow/coldstart.yaml", "parm/wflow/post.yaml", "parm/wflow/plot.yaml"]|include }}'
```
Is ther reason why plot task needed this because we can't put a task entry in the xml file unless it is used now?
I think the lack of jinja templating in the yaml template files (not the support we have inside dictionary entries but for a "block" of code) is restrictive compared to the xml template. In my opinion, specifying a list of taskgroups in the user config files looks cumbersome for me compared having flags to test if the task should be run.

christinaholtNOAA replied:
I'm not sure I understand your concern. The taskgroups entry specifies which tasks to run. There is not a mechanism currently to "add to the base tasks" other than just define the whole list of groups of tasks that should run. 

In  say that this is far less restrictive in the sense that we can literally define any task to run with this method, not just those already enabled by pre-defined templates.

danielabdi-noaa replied:
I was talking about the lack of jinja templating in the workflow yaml files under `param/wflow` such as `default_workflow.yaml` being restrictive. I am not sure if this is be design, for example, to avoid the workflow yaml files from going down the path of the previous approach of using jinja templates and becoming unreadable.
On that note, I need your feedback on this commit where I added the rrfs prep tasks, assuming jinja templating is implemented in the workflow yaml files. 
https://github.com/ufs-community/ufs-srweather-app/pull/647/commits/a6bde3447e9f4a8d572e27001dcc7775e6952f8b

With regard to defining `taskgroups` in the user config files is what I refered to as cumbersome. I think it would be easier for the user to just specify whether to run a task or not using the old `workflow_switches:` set of variables or other logical variables that enable a whole set of tasks. Depending on that, we can figure out what workflow yaml files to include and compose the whole workflow hidden from the user.

danielabdi-noaa commented on tests/WE2E/test_configs/grids_extrn_mdls_suites_community/config.grid_RRFS_NA_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta.yaml:
```diff
@@ -23,15 +23,21 @@ task_get_extrn_lbcs:
+        walltime: 01:00:00
+    metatask_run_ens_post:
+      metatask_run_post:
+        task_run_post_mem#mem#_f#fhr#:
```
Does non-ensemble runs also have a mem suffix now?

christinaholtNOAA replied:
That's correct. A deterministic run will now be "mem000" for consistency and simplicity across quite a bit of logic.

christinaholtNOAA committed:
Changes needed for AQM Tests

At this point, the workflow runs through to nexus_emission tasks, where
those fail becuase they run out of memory.

mkavulich reviewed (approved):
Compared comprehensive test results on Hera. Found bit-for-bit identical .nc files between develop and this PR branch. Only failures in each were 3km GFSv16 physics cases, which are known failures per #693

Here are the results for the curious:
 - `develop` /scratch2/BMC/fv3lam/kavulich/UFS/workdir/test_develop/2023-03-24/expt_dirs/73321889
 - `christinaholtNOAA:rocoto_yaml` /scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_676/expt_dirs/156882be/

mkavulich reviewed (changes requested):
Sorry for the whiplash but I have to take back my approval: I forgot that verification tasks were not yet included in the comprehensive suite. I discovered some errors when attempting to launch verification WE2E tests: The `MET_ensemble_verification` test ran successfully, but the other two failed to even launch.

For test `MET_verification`:

```
  ========================================================================
  Starting experiment generation...
  ========================================================================

  ========================================================================
  Starting function setup() in "setup.py"...
  ========================================================================
Found and allowing key metatask_run_ensemble
  WORKFLOW ID = id_1680036716

  The ozone parameter used for this experiment is ozphys_2015.


  The list that sets the mapping between symlinks in the cycle
  directory, and the files in the FIXam directory has been updated
  to include the ozone production/loss file.


        CYCLEDIR_LINKS_TO_FIXam_FILES_MAPPING = ( \
"aerosol.dat                | global_climaeropac_global.txt" \
"co2historicaldata_2010.txt | fix_co2_proj/global_co2historicaldata_2010.txt" \
"co2historicaldata_2011.txt | fix_co2_proj/global_co2historicaldata_2011.txt" \
"co2historicaldata_2012.txt | fix_co2_proj/global_co2historicaldata_2012.txt" \
"co2historicaldata_2013.txt | fix_co2_proj/global_co2historicaldata_2013.txt" \
"co2historicaldata_2014.txt | fix_co2_proj/global_co2historicaldata_2014.txt" \
"co2historicaldata_2015.txt | fix_co2_proj/global_co2historicaldata_2015.txt" \
"co2historicaldata_2016.txt | fix_co2_proj/global_co2historicaldata_2016.txt" \
"co2historicaldata_2017.txt | fix_co2_proj/global_co2historicaldata_2017.txt" \
"co2historicaldata_2018.txt | fix_co2_proj/global_co2historicaldata_2018.txt" \
"co2historicaldata_2019.txt | fix_co2_proj/global_co2historicaldata_2019.txt" \
"co2historicaldata_2020.txt | fix_co2_proj/global_co2historicaldata_2020.txt" \
"co2historicaldata_2021.txt | fix_co2_proj/global_co2historicaldata_2021.txt" \
"co2historicaldata_glob.txt | global_co2historicaldata_glob.txt" \
"co2monthlycyc.txt          | co2monthlycyc.txt" \
"global_h2oprdlos.f77       | global_h2o_pltc.f77" \
"global_albedo4.1x1.grb     | global_albedo4.1x1.grb" \
"global_zorclim.1x1.grb     | global_zorclim.1x1.grb" \
"global_tg3clim.2.6x1.5.grb | global_tg3clim.2.6x1.5.grb" \
"sfc_emissivity_idx.txt     | global_sfc_emissivity_idx.txt" \
"solarconstant_noaa_an.txt  | global_solarconstant_noaa_an.txt" \
"global_o3prdlos.f77        | ozprdlos_2015_new_sbuvO3_tclm15_nuchem.f77" \
)
        

*********************************************************************
FATAL ERROR:
Experiment generation failed. See the error message(s) printed below.
For more detailed information, check the log file from the workflow
generation script: log.run_WE2E_tests
*********************************************************************

Traceback (most recent call last):
  File "/scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_676/ufs-srweather-app/tests/WE2E/./run_WE2E_tests.py", line 493, in <module>
    run_we2e_tests(homedir,args)
  File "/scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_676/ufs-srweather-app/tests/WE2E/./run_WE2E_tests.py", line 199, in run_we2e_tests
    expt_dir = generate_FV3LAM_wflow(ushdir,logfile=f"{ushdir}/log.generate_FV3LAM_wflow",
  File "/scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_676/ufs-srweather-app/tests/WE2E/../../ush/generate_FV3LAM_wflow.py", line 71, in generate_FV3LAM_wflow
    expt_config = setup(ushdir,debug=debug)
  File "/scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_676/ufs-srweather-app/tests/WE2E/../../ush/setup.py", line 1231, in setup
    raise Exception(
Exception: 
            Ensemble verification can not be run unless running in ensemble mode:
               DO_ENSEMBLE = "False"
               Ensemble verification tasks: metatask_vx_ens_member
```

For test `MET_verification_only_vx`

```
  ========================================================================
  Starting function setup() in "setup.py"...
  ========================================================================
Found and allowing key task_get_obs_ccpa
Found and allowing key task_get_obs_mrms
Found and allowing key task_get_obs_ndas

*********************************************************************
FATAL ERROR:
Experiment generation failed. See the error message(s) printed below.
For more detailed information, check the log file from the workflow
generation script: log.run_WE2E_tests
*********************************************************************

Traceback (most recent call last):
  File "/scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_676/ufs-srweather-app/tests/WE2E/./run_WE2E_tests.py", line 493, in <module>
    run_we2e_tests(homedir,args)
  File "/scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_676/ufs-srweather-app/tests/WE2E/./run_WE2E_tests.py", line 199, in run_we2e_tests
    expt_dir = generate_FV3LAM_wflow(ushdir,logfile=f"{ushdir}/log.generate_FV3LAM_wflow",
  File "/scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_676/ufs-srweather-app/tests/WE2E/../../ush/generate_FV3LAM_wflow.py", line 71, in generate_FV3LAM_wflow
    expt_config = setup(ushdir,debug=debug)
  File "/scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_676/ufs-srweather-app/tests/WE2E/../../ush/setup.py", line 380, in setup
    expt_config = load_config_for_setup(USHdir, default_config_fp, user_config_fp)
  File "/scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_676/ufs-srweather-app/tests/WE2E/../../ush/setup.py", line 106, in load_config_for_setup
    raise Exception(errmsg)
Exception: Invalid key(s) specified in /scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_676/ufs-srweather-app/ush/config.yaml:
RUN_TASKS_METVX_DET = True

Check /scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_676/ufs-srweather-app/ush/config_defaults.yaml for allowed user-specified variables
```

mkavulich committed:
Remove old invalid variable from VX only test

mkavulich committed:
Add final check for invalid tasks and metatasks before writing rocoto yaml file

christinaholtNOAA committed:
All fixes from review comments, test failures.

christinaholtNOAA committed:
Undo aqm community overwrite.

danielabdi-noaa commented on parm/wflow/default_workflow.yaml:
```diff
@@ -0,0 +1,44 @@
+    JOBSdir: '{{ user.JOBSdir }}'
+    LOAD_MODULES_RUN_TASK_FP: '{{ workflow.LOAD_MODULES_RUN_TASK_FP }}'
+    LOGDIR: '{% if user.RUN_ENVIR == "nco" %}{{"{}/@Y@m@d".format(nco.LOGBASEDIR)}}{% else %}{{nco.LOGBASEDIR }}{% endif %}'
+    LOGEXT: '{% if user.RUN_ENVIR == "nco" %}{{".{}.log".format(workflow.WORKFLOW_ID)}}{% else %}{{".log"}}{% endif %}'
```
This is defined but not used in the logfiles of tasks i.e. the `join:` entries. The hardcoded `.log` should be replaced by `&LOGEXT;` so that log files in NCO mode will have the workflow IDs attached to them to avoid conflict.

christinaholtNOAA replied:
Gotcha! Nice catch. I'm putting that back in for this (hopefully) last round of tests.

danielabdi-noaa commented on parm/wflow/plot.yaml:
```diff
@@ -0,0 +1,49 @@
+task_plot_allvars:
+  <<: *default_task
+  command: '&LOAD_MODULES_RUN_TASK_FP; "plot_allvars" "&JOBSdir;/JREGIONAL_MAKE_GRID"'
+  join: !cycstr '&LOGDIR;/{{ jobname }}_@Y@m@d@H.log'
```
Here for example.

christinaholtNOAA committed:
Set all the paths the same way.

mkavulich committed:
Merge remote-tracking branch 'upstream/develop' into rocoto_yaml

Several merge conflicts manually resolved

mkavulich committed:
Post-merge cleanup to get WE2E tests to run (still a couple failures)

christinaholtNOAA committed:
Addressing last of the failures and comments.

christinaholtNOAA committed:
Addressing error in removing the memory tag.

mkavulich reviewed (approved):
Verification failures have been addressed. All comprehensive tests now pass on Hera. This PR can be merged.

christinaholtNOAA commented:
The pipelines seem stuck this evening while monitoring them on Jenkins. I saw the failure on Gaea and pushed the change that should fix it, but didn't see a way to stop the current workflow and start a new one.

mkavulich commented:
Note on testing: this PR appears to fix #688 as well.

christinaholtNOAA committed:
Remove partition flags if set to none.

mkavulich committed:
Better solution that doesn't require empty partition entries for machine files

MichaelLueken commented:
@christinaholtNOAA After resubmitting the Jenkins tests this morning, I'm seeing the following-

Cheyenne:
```
03/31/23 08:53:28 MDT :: FV3LAM_wflow.xml :: qsub: Illegal attribute or resource value
03/31/23 08:53:30 MDT :: FV3LAM_wflow.xml :: qsub: Illegal attribute or resource value
Reading database for experiment grid_CONUS_25km_GFDLgrid_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16, updating experiment dictionary
03/31/23 08:53:35 MDT :: FV3LAM_wflow.xml :: qsub: Illegal attribute or resource value
03/31/23 08:53:37 MDT :: FV3LAM_wflow.xml :: qsub: Illegal attribute or resource value
Reading database for experiment grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_HRRR, updating experiment dictionary
03/31/23 08:53:40 MDT :: FV3LAM_wflow.xml :: qsub: Illegal attribute or resource value
03/31/23 08:53:42 MDT :: FV3LAM_wflow.xml :: qsub: Illegal attribute or resource value
Reading database for experiment grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta, updating experiment dictionary
03/31/23 08:53:45 MDT :: FV3LAM_wflow.xml :: qsub: Illegal attribute or resource value
03/31/23 08:53:47 MDT :: FV3LAM_wflow.xml :: qsub: Illegal attribute or resource value
Reading database for experiment grid_RRFS_CONUScompact_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16, updating experiment dictionary
03/31/23 08:53:50 MDT :: FV3LAM_wflow.xml :: qsub: Illegal attribute or resource value
03/31/23 08:53:53 MDT :: FV3LAM_wflow.xml :: qsub: Illegal attribute or resource value
Reading database for experiment grid_RRFS_NA_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta, updating experiment dictionary
Experiment grid_CONUS_25km_GFDLgrid_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16 status is SUBMITTING
Experiment grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_HRRR status is SUBMITTING
Experiment grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta status is SUBMITTING
Experiment grid_RRFS_CONUScompact_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16 status is SUBMITTING
Experiment grid_RRFS_NA_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta status is SUBMITTING
```

Gaea:
```
03/31/23 10:52:53 EDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:52:54 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:52:59 EDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:53:00 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
Reading database for experiment grid_RRFS_CONUScompact_13km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta, updating experiment dictionary
03/31/23 10:53:06 EDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:53:07 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:53:14 EDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:53:15 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
Reading database for experiment grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta, updating experiment dictionary
03/31/23 10:53:20 EDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:53:22 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:53:28 EDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:53:29 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
Reading database for experiment grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2, updating experiment dictionary
03/31/23 10:53:35 EDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:53:36 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:53:42 EDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:53:44 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
Reading database for experiment grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16, updating experiment dictionary
03/31/23 10:53:51 EDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:53:53 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:53:59 EDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:01 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
Reading database for experiment grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_HRRR, updating experiment dictionary
03/31/23 10:54:07 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:13 EDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:15 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
Reading database for experiment grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR, updating experiment dictionary
03/31/23 10:54:22 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:22 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:23 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:23 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:24 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:25 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:25 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:26 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:26 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem001 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:27 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem002 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:27 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem001 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:28 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem002 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:29 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem001 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:29 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem002 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:30 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem001 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:30 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem002 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:37 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:38 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:38 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:38 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:39 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:39 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:41 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:41 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:42 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem001 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:43 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem002 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:43 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem001 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:44 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem002 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:44 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem001 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:45 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem002 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:46 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem001 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:54:46 EDT :: FV3LAM_wflow.xml :: Submission of run_fcst_mem002 failed!  sbatch: error: Batch job submission failed: Node count specification invalid
Reading database for experiment nco_ensemble, updating experiment dictionary
03/31/23 10:54:54 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:55:00 EDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 10:55:02 EDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
Reading database for experiment grid_SUBCONUS_Ind_3km_ics_RAP_lbcs_RAP_suite_RRFS_v1beta, updating experiment dictionary
Experiment grid_RRFS_CONUScompact_13km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta status is SUBMITTING
Experiment grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta status is SUBMITTING
Experiment grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2 status is SUBMITTING
Experiment grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16 status is SUBMITTING
Experiment grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_HRRR status is SUBMITTING
Experiment grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR status is SUBMITTING
Experiment nco_ensemble status is SUBMITTING
Experiment grid_SUBCONUS_Ind_3km_ics_RAP_lbcs_RAP_suite_RRFS_v1beta status is SUBMITTING
```

Orion:
```
03/31/23 09:57:07 CDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 09:57:07 CDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 09:57:08 CDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 09:57:08 CDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
Reading database for experiment grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta, updating experiment dictionary
03/31/23 09:57:09 CDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 09:57:10 CDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 09:57:10 CDT :: FV3LAM_wflow.xml :: sbatch: error: Batch job submission failed: Node count specification invalid
03/31/23 09:57:11 CDT :: FV3LAM_wflow.xml :: WARNING: job submission failed: sbatch: error: Batch job submission failed: Node count specification invalid
Reading database for experiment grid_SUBCONUS_Ind_3km_ics_RAP_lbcs_RAP_suite_RRFS_v1beta, updating experiment dictionary
Experiment grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta status is SUBMITTING
Experiment grid_SUBCONUS_Ind_3km_ics_RAP_lbcs_RAP_suite_RRFS_v1beta status is SUBMITTING
```

It looks like the issue from yesterday is still occurring within the Jenkins pipeline.  I'll look into trying to manually run the Jenkins pipelines and see if anything jumps out as an issue.

christinaholtNOAA committed:
Attempt to fix 0 cores for forecast

christinaholtNOAA committed:
Fix bad name.

christinaholtNOAA commented:
@mkavulich and I are doing some rapid dev/test iterations to try to fix these Cheyenne issues. I _think_ they should also take care of the issues on other platforms.

mkavulich commented:
Ran manual fundamental tests on Cheyenne and Hera, all expected tests passed: Note that there is a known failure in develop for test grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR in the fundamental test on Hera, see #688 for more details.

@MichaelLueken I think we're ready to kick off Jenkins tests again.

mkavulich commented:
@MichaelLueken I spoke too soon, the final verification tasks are failing on the one Cheyenne verification test. Investigating now.

mkavulich commented:
Okay, false alarm, the failure was due to running out of disk space. @MichaelLueken Please start up the Jenkins tests

MichaelLueken commented:
@mkavulich and @christinaholtNOAA The Jenkins tests have been resubmitted.  I have also run the Jenkins fundamental tests manually on Jet and all tests successfully passed, so I'm feeling positive on this batch of tests.  Thanks!

MichaelLueken commented:
@danielabdi-noaa Since you had left several comments in this PR, I'd like to check and make sure that you are okay with these changes before moving forward.  The Jenkins tests are still running on Gaea, but I would like to ensure that all parties are okay now.  Thank you very much for your time!

danielabdi-noaa reviewed (approved):
@MichaelLueken Yes, my suggestions have been more or less addressed so approving it now.

christinaholtNOAA commented:
@MichaelLueken Can we merge this PR while the gaea test are still stuck? I'd like to unblock the other PRs and address any remaining gaea issues next week, separately, if they exist.

MichaelLueken commented:
@christinaholtNOAA Sure.  Since the manual tests on Hera and Jet passed and the Jenkins tests on Cheyenne and Orion have successfully passed, I think we can move forward with these changes now and address Gaea issues on Monday.  Have a great weekend!
